### DISCUSSION TITLE
List of possible Event Listeners

### DISCUSSION THREAD

**Discussion by smezger123** [matter-js/matter.js#1555] [Q&A] [answered]
Hi community,

I have already spent many hours using the online help and search engines. But I cannot find a list of all possible event handlers of an endpoint.

My very simple problem: I want to intercept every switch request for a simple on-off device.

endpoint1.events.onOff.onOff$Changed.on(value => ....
is only triggered when the state changes. For example switching from on to off. And that´s working fine!

But I also want to react to the event when the device is already on and another on request still comes in.
For example:
endpoint1.events.onOff.onOff$**Always**.on(value => ...

But of course, "always" doesn't exist. What else can I do?

Thank you very much.

**Comment by lauckhart** [comment, marked as answer]
The full set of events for a behavior is:

* `attr$Changing` and `attr$Changed` for for enabled attribute in the corresponding cluster definition
* `event` for each enabled event in the cluster
* additional events defined in the server implementation

We have plans to generate proper documentation. Unfortunately there aren't tools that will do it automatically so it will take a bit of work.

Regarding "onOff$Always", as you note the attribute events only trigger when attributes mutate. If you want to detect "on" or "off" commands regardless of current state then you probably want to extend OnOffServer, then override `on` and/or `off` to add your custom logic.

If you are looking for a generic way to do this for any command the answer is a bit more involved. If you're trying to do that give us a bit more background on your use case and we can give you additional suggestions.

**Reply by Apollon77** [reply]
BTW: This is also documented in https://github.com/project-chip/matter.js/blob/main/docs/MIGRATION_GUIDE_08.md#react-to-change-events-on-cluster-attributes

And yes currently there is no possibility to get an event for "value got updated aka "Unchanged" other then overwriting the command handlers and extending on this level, but that's a bit more enhanced usecase as soon as it is more then on/off :-)

**Comment by smezger123** [comment]
Hi lauckhart,

thank you very much for your reply.
I have now implemented the hint from your post with the OnOffServer class. It does what it is supposed to do. Thanks!

```
    class MyEvent_RobotCleaner extends OnOffLightRequirements.OnOffServer { on() { this.task("start"); } off() { this.task("home"); } task(value) { executeRobotCleanerHTTPCommand(value); } }
    await aggregator.add(new Endpoint(OnOffLightDevice.with(BridgedDeviceBasicInformationServer).with(MyEvent_RobotCleaner), { id: "RobotCleaner", bridgedDeviceBasicInformation: {nodeLabel: "vacuum-cleaner", serialNumber: "prd-endpoint3"} }));
```


Background: I only need it to convert voice commands from Alexa to HTTP REST calls to my SmartHome. Local fulfillment without cloud peerings.

Current example: I want to turn on the "light" with name vacuum-cleaner. After the vacuum cleaner has finished his job, the "light" in Matter is still in state "on". Next time I still want to turn the "light" on again without having to turn it off first.
Implementing a status query to the vacuum cleaner so that the "light" has the same status would be too much work and is unnecessary. Also because I have a lot of these "lights"...
